mcap merge fails on missing schema

**Description**
When merging 2 mcaps with the CLI tool, and one of them has a schema ID of 0 (no schema, just 1 channel with schema ID 0 and 1 message on that channel), the CLI tool crashes with:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1006df838]

goroutine 1 [running]:
github.com/foxglove/mcap/go/cli/mcap/cmd.(*mcapMerger).addSchema(0x14003405d48, 0x0?, 0x1c1, 0x0)
        /private/tmp/mcap-20230216-70145-8qbk2v/mcap-releases-mcap-cli-v0.0.25/go/cli/mcap/cmd/merge.go:109 +0x28
github.com/foxglove/mcap/go/cli/mcap/cmd.(*mcapMerger).mergeInputs(0x14003405d48, {0x100acecc0, 0x14000011908}, {0x140004d0000, 0x1c4, 0x0?})
        /private/tmp/mcap-20230216-70145-8qbk2v/mcap-releases-mcap-cli-v0.0.25/go/cli/mcap/cmd/merge.go:186 +0x6e4
github.com/foxglove/mcap/go/cli/mcap/cmd.glob..func12(0x100ff6060?, {0x140004ae000?, 0x1c4, 0x200?})
        /private/tmp/mcap-20230216-70145-8qbk2v/mcap-releases-mcap-cli-v0.0.25/go/cli/mcap/cmd/merge.go:279 +0x29c
github.com/spf13/cobra.(*Command).execute(0x100ff6060, {0x140004a6000, 0x1c6, 0x200})
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:876 +0x4b8
github.com/spf13/cobra.(*Command).ExecuteC(0x100ff5de0)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:990 +0x354
github.com/spf13/cobra.(*Command).Execute(...)
        /Users/brew/Library/Caches/Homebrew/go_mod_cache/pkg/mod/github.com/spf13/cobra@v1.5.0/command.go:918
github.com/foxglove/mcap/go/cli/mcap/cmd.Execute()
        /private/tmp/mcap-20230216-70145-8qbk2v/mcap-releases-mcap-cli-v0.0.25/go/cli/mcap/cmd/root.go:21 +0x28
main.main()
        /private/tmp/mcap-20230216-70145-8qbk2v/mcap-releases-mcap-cli-v0.0.25/go/cli/mcap/main.go:6 +0x1c
```

- Version: `mcap version v0.0.25`
- Platform: macOS Monterey 12.5.1

**Steps To Reproduce**
<!--- Include the minimum possible code or example data to reproduce the problem. -->
* Create an mcap with the python API with 1 channel set to schema ID 0 and 1 or more messages
* Run the merge command with `mcap merge no-schema.mcap regular.mcap`
* Tool crashes with runtime error above

**Expected Behavior**
Tool should generate merged mcap so that it can be loaded into Foxglove studio (ref: [#5351 Support schemaless channels](https://github.com/foxglove/studio/pull/5351))

--- a/go/cli/mcap/cmd/merge.go
+++ b/go/cli/mcap/cmd/merge.go
@@ -75,6 +75,9 @@ func (m *mcapMerger) outputChannelID(inputID int, inputChannelID uint16) (uint16
 }
 
 func (m *mcapMerger) outputSchemaID(inputID int, inputSchemaID uint16) (uint16, bool) {
+	if inputSchemaID == 0 {
+		return 0, true
+	}
 	v, ok := m.schemaIDs[schemaID{
 		inputID:  inputID,
 		schemaID: inputSchemaID,
@@ -183,9 +186,11 @@ func (m *mcapMerger) mergeInputs(w io.Writer, inputs []io.Reader) error {
 			}
 			return fmt.Errorf("failed to read first message on input %d: %w", inputID, err)
 		}
-		schema.ID, err = m.addSchema(writer, inputID, schema)
-		if err != nil {
-			return fmt.Errorf("failed to add initial schema for input %d: %w", inputID, err)
+		if schema != nil {
+			schema.ID, err = m.addSchema(writer, inputID, schema)
+			if err != nil {
+				return fmt.Errorf("failed to add initial schema for input %d: %w", inputID, err)
+			}
 		}
 		message.ChannelID, err = m.addChannel(writer, inputID, channel)
 		if err != nil {

--- a/go/cli/mcap/cmd/merge_test.go
+++ b/go/cli/mcap/cmd/merge_test.go
@@ -114,11 +114,15 @@ func TestSchemalessChannelInput(t *testing.T) {
 	it, err := reader.Messages(readopts.UsingIndex(false))
 	assert.Nil(t, err)
 	messages := make(map[string]int)
+	schemaIDs := make(map[uint16]int)
 	err = mcap.Range(it, func(schema *mcap.Schema, channel *mcap.Channel, message *mcap.Message) error {
 		messages[channel.Topic]++
+		schemaIDs[channel.SchemaID]++
 		return nil
 	})
 	assert.Nil(t, err)
 	assert.Equal(t, 100, messages["/foo"])
 	assert.Equal(t, 100, messages["/bar"])
+	assert.Equal(t, 100, schemaIDs[0])
+	assert.Equal(t, 100, schemaIDs[1])
 }

--- a/go/cli/mcap/cmd/merge_test.go
+++ b/go/cli/mcap/cmd/merge_test.go
@@ -17,9 +17,11 @@ func prepInput(t *testing.T, w io.Writer, schemaID uint16, channelID uint16, top
 	assert.Nil(t, err)
 
 	assert.Nil(t, writer.WriteHeader(&mcap.Header{Profile: "testprofile"}))
-	assert.Nil(t, writer.WriteSchema(&mcap.Schema{
-		ID: schemaID,
-	}))
+	if schemaID != 0 {
+		assert.Nil(t, writer.WriteSchema(&mcap.Schema{
+			ID: schemaID,
+		}))
+	}
 	assert.Nil(t, writer.WriteChannel(&mcap.Channel{
 		ID:       channelID,
 		SchemaID: schemaID,
@@ -95,4 +97,28 @@ func TestMultiChannelInput(t *testing.T) {
 	assert.Equal(t, 100, messages["/foo"])
 	assert.Equal(t, 100, messages["/bar"])
 	assert.Equal(t, 100, messages["/baz"])
+}
+func TestSchemalessChannelInput(t *testing.T) {
+	buf1 := &bytes.Buffer{}
+	buf2 := &bytes.Buffer{}
+	prepInput(t, buf1, 0, 1, "/foo")
+	prepInput(t, buf2, 1, 1, "/bar")
+	merger := newMCAPMerger(mergeOpts{})
+	output := &bytes.Buffer{}
+	assert.Nil(t, merger.mergeInputs(output, []io.Reader{buf1, buf2}))
+
+	// output should now be a well-formed mcap
+	reader, err := mcap.NewReader(output)
+	assert.Nil(t, err)
+	assert.Equal(t, reader.Header().Profile, "testprofile")
+	it, err := reader.Messages(readopts.UsingIndex(false))
+	assert.Nil(t, err)
+	messages := make(map[string]int)
+	err = mcap.Range(it, func(schema *mcap.Schema, channel *mcap.Channel, message *mcap.Message) error {
+		messages[channel.Topic]++
+		return nil
+	})
+	assert.Nil(t, err)
+	assert.Equal(t, 100, messages["/foo"])
+	assert.Equal(t, 100, messages["/bar"])
 }

